src/main/java/frc/robot/subsystems/drive/Drive.java

```diff
@@ -0,0 +1,227 @@
+{
+    private static Drive _instance;
+
+    public static Drive getInstance()
+    {
+        if (_instance == null)
+        {
+            _instance = new Drive();
+        }
+
+        return _instance;
+    }
+
+    //private AHRS           _gyro; // navX2
+    private ADIS16448_IMU  _gyro;
+    private double         _gyroOffset; // angle to adjust gyro by for zeroing
+
+    private Vector         _origin; // origin of rotation relative to the center of the robot frame
+
+    private SwerveModule[] _swerveModules; // all of the swerve modules
+
+    // odometry
+    private double         _rotationHeading; // used to calculate angular velocity
+    private double         _rotationVelocity; // how fast are we turning?
+
+    private Vector         _position; // current odometer reading
+    private Vector         _velocity; // change in odometer reading
+
+    private Drive()
+    {
+        _origin             = new Vector();
+
+        //_gyro = new AHRS(SPI.Port.kMXP);
+        _gyro = new ADIS16448_IMU(IMUAxis.kX, SPI.Port.kMXP, CalibrationTime._4s);
+
+        SwerveModule fl = new SwerveModule(-Constants.Drive.BASE_WIDTH / 2,  Constants.Drive.BASE_LENGTH / 2, 37.5, 0.0, 1, 2, 0);// FIXME
+        SwerveModule fr = new SwerveModule( Constants.Drive.BASE_WIDTH / 2,  Constants.Drive.BASE_LENGTH / 2, -150.2, 0.0, 5, 6, 3);
+        SwerveModule bl = new SwerveModule(-Constants.Drive.BASE_WIDTH / 2, -Constants.Drive.BASE_LENGTH / 2, -158.2, 180.0, 3, 4, 1);
+        SwerveModule br = new SwerveModule( Constants.Drive.BASE_WIDTH / 2, -Constants.Drive.BASE_LENGTH / 2, -53.3, 180.0, 7, 8, 2);
+
+        _swerveModules = new SwerveModule[]// turn the swerve modules into an array, so they can be easily accessed, with an arbitrary number of them
+        { 
+            fl, 
+            fr,
+            bl,
+            br 
+        };
+
+        setOrigin(0, 0);// rotate around the center of the robot, by default
+        resetOdometer();// initialize the odometer to 0, 0
+    }
+
+    public void chassisDrive(double drive, double strafe, double rotate)
+    {
+        Vector translateVector = new Vector(strafe, drive);
+        
+        drive(translateVector, rotate);
+    }
+
+    public void fieldDrive(double x, double y, double rotate)
+    {
+        Vector translateVector = new Vector(x, y);
+
+        translateVector.translatePolarPosition(0, -getHeading());
+        
+        drive(translateVector, rotate);
+    }
+
+    private void drive(Vector translateVector, double rotate)
+    {
+        Vector[] moduleCommands = new Vector[_swerveModules.length];
+
+        double maxSpeed = 0;
+
+        for (int i = 0; i < _swerveModules.length; i++)
+        {
+            Vector modulePosition = _swerveModules[i].subtract(_origin);
+            Vector rotateVector = new Vector(modulePosition.getY(), -modulePosition.getX());// clockwise 90 deg.
+            rotateVector = rotateVector.multiply(rotate / Constants.Drive.TYPICAL_MODULE_DIST);
```

Comment:
Do we want to use module-specific scalars to go from rotational velocity to rotational percent output?

Reply:
Same question